File: packages/php/src/Types/Metadata.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types;

/**
 * Document metadata.
 *
 * All fields are optional and depend on the file format, extraction configuration,
 * and postprocessors enabled.
 *
 * @property-read string|null $language Document language (ISO 639-1 code)
 * @property-read string|null $date Document date (ISO 8601 format)
 * @property-read string|null $subject Document subject
 * @property-read string|null $formatType Format discriminator ("pdf", "excel", "email", etc.)
 * @property-read string|null $title Document title
 * @property-read array<string>|null $authors Document authors
 * @property-read array<string>|null $keywords Document keywords
 * @property-read string|null $createdAt Creation date (ISO 8601)
 * @property-read string|null $modifiedAt Modification date (ISO 8601)
 * @property-read string|null $createdBy Creator/application name
 * @property-read string|null $producer Producer/generator
 * @property-read int|null $pageCount Number of pages
 * @property-read array<string, mixed> $custom Additional custom metadata from postprocessors
 */
readonly class Metadata
{
    /**
     * @param array<string>|null $authors
     * @param array<string>|null $keywords
     * @param array<string, mixed> $custom
     */
    public function __construct(
        public ?string $language = null,
        public ?string $date = null,
        public ?string $subject = null,
        public ?string $formatType = null,
        public ?string $title = null,
        public ?array $authors = null,
        public ?array $keywords = null,
        public ?string $createdAt = null,
        public ?string $modifiedAt = null,
        public ?string $createdBy = null,
        public ?string $producer = null,
        public ?int $pageCount = null,
        public array $custom = [],
    ) {
    }

    /**
     * Create Metadata from array returned by extension.
     *
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        $knownFields = [
            'language',
            'date',
            'subject',
            'format_type',
            'title',
            'authors',
            'keywords',
            'created_at',
            'modified_at',
            'created_by',
            'producer',
            'page_count',
        ];

        /** @var string|null $language */
        $language = $data['language'] ?? null;

        /** @var string|null $date */
        $date = $data['date'] ?? null;

        /** @var string|null $subject */
        $subject = $data['subject'] ?? null;

        /** @var string|null $formatType */
        $formatType = $data['format_type'] ?? null;

        /** @var string|null $title */
        $title = $data['title'] ?? null;

        /** @var array<string>|null $authors */
        $authors = $data['authors'] ?? null;

        /** @var array<string>|null $keywords */
        $keywords = $data['keywords'] ?? null;

        /** @var string|null $createdAt */
        $createdAt = $data['created_at'] ?? null;

        /** @var string|null $modifiedAt */
        $modifiedAt = $data['modified_at'] ?? null;

        /** @var string|null $createdBy */
        $createdBy = $data['created_by'] ?? null;

        /** @var string|null $producer */
        $producer = $data['producer'] ?? null;

        /** @var int|null $pageCount */
        $pageCount = $data['page_count'] ?? null;

        $custom = [];
        foreach ($data as $key => $value) {
            if (!in_array($key, $knownFields, true)) {
                $custom[$key] = $value;
            }
        }

        return new self(
            language: $language,
            date: $date,
            subject: $subject,
            formatType: $formatType,
            title: $title,
            authors: $authors,
            keywords: $keywords,
            createdAt: $createdAt,
            modifiedAt: $modifiedAt,
            createdBy: $createdBy,
            producer: $producer,
            pageCount: $pageCount,
            custom: $custom,
        );
    }

    /**
     * Get a custom metadata field.
     *
     * @param string $key
     * @return mixed
     */
    public function getCustom(string $key): mixed
    {
        return $this->custom[$key] ?? null;
    }

    /**
     * Check if a custom metadata field exists.
     */
    public function hasCustom(string $key): bool
    {
        return isset($this->custom[$key]);
    }
}

File: crates/xberg/src/pdf/metadata.rs
//! PDF metadata types (backend-agnostic).
//!
//! These types are shared between the oxide extraction backend and any callers
//! that consume PDF metadata. No PDF-library-specific dependencies.

use serde::{Deserialize, Serialize};

use crate::types::PageStructure;

/// PDF-specific metadata.
///
/// Contains metadata fields specific to PDF documents that are not in the common
/// `Metadata` structure. Common fields like title, authors, keywords, and dates
/// are at the `Metadata` level.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[cfg_attr(feature = "api", derive(utoipa::ToSchema))]
pub struct PdfMetadata {
    /// PDF version (e.g., "1.7", "2.0")
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pdf_version: Option<String>,

    /// PDF producer (application that created the PDF)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub producer: Option<String>,

    /// Whether the PDF is encrypted/password-protected
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_encrypted: Option<bool>,

    /// First page width in points (1/72 inch)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub width: Option<i64>,

    /// First page height in points (1/72 inch)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub height: Option<i64>,

    /// Total number of pages in the PDF document
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page_count: Option<u32>,

    /// How strongly the document's most scan-like page resembles a scan, in `[0.0, 1.0]`.
    ///
    /// `None` when the document could not be inspected. A full-page raster with no
    /// visible text scores at least `0.85`; a born-digital slide with a full-bleed
    /// background image scores `0.50`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scanned_confidence: Option<f32>,

    /// Pages that look like scans (1-indexed), using the default confidence threshold.
    ///
    /// `None` when the document could not be inspected; empty when no page qualifies.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scanned_pages: Option<Vec<u32>>,
}
/// Complete PDF extraction metadata including common and PDF-specific fields.
///
/// Combines common document fields (title, authors, dates) with PDF-specific
/// metadata and optional page structure information.
#[cfg_attr(alef, alef(skip))]
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PdfExtractionMetadata {
    /// Document title
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,

    /// Document subject or description
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subject: Option<String>,

    /// Document authors (parsed from PDF Author field)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub authors: Option<Vec<String>>,

    /// Document keywords (parsed from PDF Keywords field)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub keywords: Option<Vec<String>>,

    /// Creation timestamp (ISO 8601 format)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub created_at: Option<String>,

    /// Last modification timestamp (ISO 8601 format)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub modified_at: Option<String>,

    /// Application or user that created the document
    #[serde(skip_serializing_if = "Option::is_none")]
    pub created_by: Option<String>,

    /// PDF-specific metadata
    pub pdf_specific: PdfMetadata,

    /// Page structure with boundaries and optional per-page metadata
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page_structure: Option<PageStructure>,
}
#[cfg_attr(alef, alef(skip))]
/// Common PDF metadata fields extracted from the document info dictionary.
///
/// Used as an intermediate type during extraction before building `PdfExtractionMetadata`.
#[cfg_attr(alef, alef(skip))]
#[derive(Debug, Clone, Default)]
pub struct CommonPdfMetadata {
    /// Document title from the PDF Info dictionary.
    pub title: Option<String>,
    /// Document subject from the PDF Info dictionary.
    pub subject: Option<String>,
    /// Document authors parsed from the PDF Author field.
    pub authors: Option<Vec<String>>,
    /// Keywords parsed from the PDF Keywords field.
    pub keywords: Option<Vec<String>>,
    /// Creation timestamp in ISO 8601 format.
    pub created_at: Option<String>,
    /// Last modification timestamp in ISO 8601 format.
    pub modified_at: Option<String>,
    /// Creator application or author name.
    pub created_by: Option<String>,
}
